cluster/images/etcd/rollback/rollback.go

```diff
@@ -0,0 +1,318 @@
+
+	// In the following, it's low level logic that saves metadata and data into v2 snapshot.
+
+	if err := os.RemoveAll(migrateDatadir); err != nil {
+		panic(err)
+	}
+	if err := os.MkdirAll(path.Join(migrateDatadir, "member", "snap"), 0700); err != nil {
+		panic(err)
+	}
+	walDir := path.Join(migrateDatadir, "member", "wal")
+
+	w, err := wal.Create(walDir, metadata)
+	if err != nil {
+		panic(err)
+	}
+	err = w.SaveSnapshot(walpb.Snapshot{Index: hardstate.Commit, Term: hardstate.Term})
+	if err != nil {
+		panic(err)
+	}
+	w.Close()
+
+	nodeIDs := []uint64{}
+	event, err := oldSt.Get(etcdserver.StoreClusterPrefix, true, false)
+	if err != nil {
+		panic(err)
+	}
+	// searching all metadata nodes and:
+	// - update store
+	// - update Nodes for ConfState
+	q := []*store.NodeExtern{}
+	q = append(q, event.Node)
+	for len(q) > 0 {
+		n := q[0]
+		q = q[1:]
+		if n.Key != etcdserver.StoreClusterPrefix {
+			v := ""
+			if !n.Dir {
+				v = *n.Value
+			}
+			if n.Key == path.Join(etcdserver.StoreClusterPrefix, "version") {
+				v = "2.3.7"
+			}
+			if _, err := st.Set(n.Key, n.Dir, v, store.TTLOptionSet{}); err != nil {
+				panic(err)
+			}
+
+			fields := strings.Split(n.Key, "/")
+			if len(fields) == 4 && fields[2] == "members" {
+				nodeID, err := strconv.ParseUint(fields[3], 16, 64)
+				if err != nil {
+					fmt.Println("wrong ID: %s", fields[3])
+					panic(err)
+				}
+				nodeIDs = append(nodeIDs, nodeID)
+			}
+		}
+		for _, next := range n.Nodes {
+			q = append(q, next)
+		}
+	}
+
+	data, err := st.Save()
+	if err != nil {
+		panic(err)
+	}
+	raftSnap := raftpb.Snapshot{
+		Data: data,
+		Metadata: raftpb.SnapshotMetadata{
+			Index: hardstate.Commit,
+			Term:  hardstate.Term,
+			ConfState: raftpb.ConfState{
+				Nodes: nodeIDs,
+			},
+		},
+	}
+	snapshotter := snap.New(path.Join(migrateDatadir, "member", "snap"))
+	if err := snapshotter.SaveSnap(raftSnap); err != nil {
+		panic(err)
+	}
+	fmt.Println("Finished.")
+}
+
+const (
+	revBytesLen            = 8 + 1 + 8
+	markedRevBytesLen      = revBytesLen + 1
+	markBytePosition       = markedRevBytesLen - 1
+	markTombstone     byte = 't'
+)
+
+func isTombstone(b []byte) bool {
+	return len(b) == markedRevBytesLen && b[markBytePosition] == markTombstone
+}
+
+func traverseAndDeleteEmptyDir(st store.Store, dir string) {
+	e, err := st.Get(dir, true, false)
+	if err != nil {
+		panic(err)
+	}
+	if len(e.Node.Nodes) == 0 {
+		st.Delete(dir, true, true)
+		return
+	}
+	for _, node := range e.Node.Nodes {
+		if !node.Dir {
+			printNode(node)
+		} else {
+			traverseAndDeleteEmptyDir(st, node.Key)
+		}
+	}
+}
+
+func printNode(node *store.NodeExtern) {
+	fmt.Printf("key:%s\n", node.Key[len("/1"):])
+}
+
+func rebuild(datadir string) ([]byte, raftpb.HardState, store.Store) {
+	waldir := path.Join(datadir, "member", "wal")
+	snapdir := path.Join(datadir, "member", "snap")
+
+	ss := snap.New(snapdir)
+	snapshot, err := ss.Load()
+	if err != nil && err != snap.ErrNoSnapshot {
+		panic(err)
```

Comment:
Similarly as above - I would really prefer to avoid panics all over the code, especially in function. Can you please return error instead here (and below)?